### ISSUE TITLE
Handle race condition between applying WAL and requesting pages from Pagers

### ISSUE THREAD

**Issue by knizhnik** [libzenith/postgres#14]
This race condition in GetPage@LSN was described in Socrates article (4.5 Secondary Compute Node).
When replica requests some page from Pager with LSN equal to the current WAL apply LSN (lets call it LSN_0), 
then there are three possible issues:
1. Pager may return page "from future" (with LSN=LSN_1 greater than specified LSN_0). It is true for Socrates: "According to the GetPage@LSN protocol, the Page Server may return a page from the future". I am not sure that our pager should follow this rule. Actually it complicates switching replica to the particular snapshot. But if such situation can happen, then smgr should wait some time, given a chance to wal receiver to apply WAL up to this LSN_1.
2. Pagers may return page with LSN-LSN_1 < LSN_0, but page was updated between LSN_1 and LSN_0. It can happen if pager didn't yet apply WAL till LSN_0. that this situation should be prevented at Pager which should suspend GetPage@LSN request until WAL is applied till this LSN. 
3. Pager returns page with LSN_0, but while waiting for this response, replica has applied WAL till LSN_2 which contains records updating this page. In this case we have three choices:
- Suspend applying WAL if it tries to update page requested from Pager (so, definitely, we should remember fact of requesting page from Pager).
- Remember WAL records touching requested page and apply them when page is received
- Resubmit GetPage@LSN request once such collision was detected.
I think that suspending WAL apply is the best approach, because it is much simpler  than trying to maintain queue of delayed WAL records and applying the for the concrete page. And there is no risk of starvation as in case of retrying GetPage request.

**Comment by knizhnik** [comment]
There are three scenarios of using GetPage@LSN:
1. Get evicted page at master
2. Get missing page at replica
3. Restore particular snapshot at replica

In first case we need to maintain  lastEvictedLsn and use it for GetPage. If we do not want to touch vanilla buffer manager, we should do it in smgr. But smgr do not know whether page is evicted from buffer manager or just saved by checkpointer. So smgr can maintain lastWrittenLsn instead of lastEvictedLsn. But it seems to be acceptable.

In case of replica, using lastEvictedLsn is not possible, because requested page may be absent at replica at all. We have to use applyWalLsn as target for GetPage.

And in the third case we just need to pass snapshot LSN to GetPage.

Looks like all this three cases can be handled by Pager in the same way, so there is no need to support more than one GetPage kinds. Pager should wait until WAL is applied till specified LSN and then found most recent version of the page with LSN <= target-LSN. Unlike Socrates, out Pager should never return "pages from future" (with LSN greater than specified).

**Comment by hlinnaka** [comment]
> 1. Pager may return page "from future" (with LSN=LSN_1 greater than specified LSN_0). It is true for Socrates: "According to the GetPage@LSN protocol, the Page Server may return a page from the future". I am not sure that our pager should follow this rule. Actually it complicates switching replica to the particular snapshot. But if such situation can happen, then smgr should wait some time, given a chance to wal receiver to apply WAL up to this LSN_1.

Yeah, returning pages "from the future" opens a big can of worms. We cannot make the kind of guarantees about index behaviour that the Socrates paper mentions. Also, what if a tuple was deleted and vacuumed away in the heap, but it should still be visible to the transaction running in the replica? Let's not allow that; our GetPage@LSN call must return the page as it is was on that exact LSN, not anything older or newer.

> 2. Pagers may return page with LSN-LSN_1 < LSN_0, but page was updated between LSN_1 and LSN_0. It can happen if pager didn't yet apply WAL till LSN_0. that this situation should be prevented at Pager which should suspend GetPage@LSN request until WAL is applied till this LSN

Agreed.

> 3. Pager returns page with LSN_0, but while waiting for this response, replica has applied WAL till LSN_2 which contains records updating this page. In this case we have three choices:
> 
> 
>     * Suspend applying WAL if it tries to update page requested from Pager (so, definitely, we should remember fact of requesting page from Pager).
> 
>     * Remember WAL records touching requested page and apply them when page is received
> 
>     * Resubmit GetPage@LSN request once such collision was detected.
> 
> I think that suspending WAL apply is the best approach, because it is much simpler  than trying to maintain queue of delayed WAL records and applying the for the concrete page. And there is no risk of starvation as in case of retrying GetPage request.

Yeah, sounds good.

**Comment by hlinnaka** [comment]
> In the first case we need to maintain lastEvictedLsn and use it for GetPage. If we do not want to touch vanilla buffer manager, we should do it in smgr. But smgr do not know whether page is evicted from buffer manager or just saved by checkpointer. So smgr can maintain lastWrittenLsn instead of lastEvictedLsn. But it seems to be acceptable.

Yeah. You could maybe distinguish checkpointer writes by looking at which process is performing the write. But I agree lastWrittenLsn should be good enough.

> In case of replica, using lastEvictedLsn is not possible, because requested page may be absent at replica at all. We have to use applyWalLsn as target for GetPage.

That's a bit unfortunate. If you have an up-to-date replica, applyWalLsn is always very close to the last flushed WAL position, so the Page Servers would need to block often because they haven't replayed the WAL yet.

Perhaps the replica could keep track of e.g. 1000 most recent blocks affected by the WAL records, and their LSNs. If you're about to request one of those most recent blocks, then use the actual LSN where that block was modified with the GetPage@LSN request. If it's not any of those blocks, then use the oldest LSN in that list. That would allow us to use a slightly older LSN in most requests, giving the Page Servers more time to apply the WAL.

But that's just an optimization, we can do it later.

**Comment by knizhnik** [comment]
I think, that main problem with replica is different and also has to be discussed.
Right now there is snadby_feedback messages delivered from replica to master because
them are not communicating directly but through safekeeper.
In principle it is possible to propagate this message though safekeeper by my intention was to completely eliinate negative influence of replicas at master (suspending autovacuum of old versions).

So if we are going to execute read-only queries at standby, we may intentionally delay applying WAL at replica. In this case applyLsn horizon at replica will be significantly smaller than at Pagers.
